src/WebJobs.Script.WebHost/Diagnostics/UserLogMetricsLoggerProvider.cs

```diff
@@ -0,0 +1,33 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using Microsoft.Azure.WebJobs.Script.Diagnostics;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost.Diagnostics
+{
+    public class UserLogMetricsLoggerProvider : ILoggerProvider, ISupportExternalScope
+    {
+        private readonly IMetricsLogger _metricsLogger;
+        private IExternalScopeProvider _scopeProvider;
+
+        public UserLogMetricsLoggerProvider(IMetricsLogger metricsLogger)
+        {
+            _metricsLogger = metricsLogger;
+        }
+
+        public ILogger CreateLogger(string categoryName)
+        {
+            return new UserLogMetricsLogger(categoryName, _metricsLogger, _scopeProvider);
```

Comment:
What is setting the _scopeProvider? This will blow up with a null ref if the `SetScopeProvider` method is not called below, right? Would it be difficult to refactor this type to move that dependency to the constructor and avoid this issue?

Reply:
The [LoggerFactory does it](https://github.com/aspnet/Logging/blob/master/src/Microsoft.Extensions.Logging/LoggerFactory.cs#L129). 

The reason you can't register one via DI and take it via ctor is because each Logger would then call `Push()`, resulting in a bunch of dupe scope objects. We do this today and I've wanted a way around it. So the LoggerFactory only calls `Push()` once for every logger that supports this external provider, knowing that they're sharing the instance they got from the LoggerFactory. I actually don't think we even need to call `Push()` ourselves inside our loggers.

I've just asked over at their repo for confirmation [here](https://github.com/aspnet/Logging/issues/875), but I believe this is why they have this pattern.

Reply:
Can we perform some validation here to ensure this has been set before initializing the `UserLogMetricsLogger`? That will give us a more descriptive error if we ever end up in this condition.

Reply:
I've added a descriptive exception there if this is null. I've also added a comment explaining that the LoggerFactory passes us the `IExternalScopeProvider`.